### PR TITLE
[BEAM-14101] [CdapIO] Add ReceiverBuilder for SparkReceiverIO

### LINKED ISSUE #21673: [CdapIO] Design and implement Spark Receiver Builder
### Context:

CDAP plugins that support streaming sources include Receiver classes (ex. [HubSpotReceiver](https://github.com/data-integrations/hubspot/blob/develop/src/main/java/io/cdap/plugin/hubspot/source/streaming/HubspotReceiver.java)) that extend `org.apache.spark.streaming.receiver.Receiver` abstract class. Receiver classes are used by plugin Streaming Utils classes (ex. [HubSpotStreamingUtils](https://github.com/data-integrations/hubspot/blob/develop/src/main/java/io/cdap/plugin/hubspot/source/streaming/HubspotStreamingUtil.java)) to provide `getStream()` method to Streaming Source classes (ex. [HubSpotStreamingSource](https://github.com/data-integrations/hubspot/blob/develop/src/main/java/io/cdap/plugin/hubspot/source/streaming/HubspotStreamingSource.java)) and usually placed in `cdap/plugin/\{<plugin name>}/plugin/source/streaming` folder (ex.[ HubSpot plugin repo](https://github.com/data-integrations/hubspot/tree/develop/src/main/java/io/cdap/plugin/hubspot/source/streaming)).

Reference information:
 * [Streaming plug-ins integration](https://docs.google.com/document/d/1T-bhd0Qk7DBePIfgHEPagYiA1oLP4z5kYEd0S1SOGxQ/edit#heading=h.o88i6p9b13o9) (Apache CDAP Connection Design Doc)

 * [Plugin integration process description](https://docs.google.com/document/d/1T-bhd0Qk7DBePIfgHEPagYiA1oLP4z5kYEd0S1SOGxQ/edit#heading=h.1h6udb1b52xc) (Apache CDAP Connection Design Doc)

 * [Streaming wrapper design](https://docs.google.com/document/d/1T-bhd0Qk7DBePIfgHEPagYiA1oLP4z5kYEd0S1SOGxQ/edit#heading=h.fcafz0ydsso1) (Apache CDAP Connection Design Doc)

### Task Description:

Required to design custom Spark receivers builder class so that the custom receivers can be used in Apache Beam connector via SparkReceiverIO interface (used in CDAP IO as a dependency).
### Acceptance criteria:

Design of builder class(es) that will create custom Spark receivers in Apache Beam connectors (`{}SparkReceiverIO{`}).
#### Note:

It is necessary that this builder class be independent of CDAP receivers and can be used by any other custom Spark receiver like a part of SparkReceiverIO.

Imported from Jira [BEAM-14101](https://issues.apache.org/jira/browse/BEAM-14101). Original Jira may contain additional context.
Reported by: elizaveta.lomteva.

### PR DESCRIPTION
**Pull request by Amar3tto**
Resolves #21673
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Amar3tto** [comment]
retest this please

**Comment by Amar3tto** [comment]
retest this please

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Amar3tto** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Amar3tto** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Lizzfox** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Amar3tto** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Lizzfox** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Amar3tto** [comment]
retest this please

**Comment by Lizzfox** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Amar3tto** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Lizzfox** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Lizzfox** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by Lizzfox** [comment]
Run Java_PVR_Flink_Docker PreCommit

**Comment by Lizzfox** [comment]
Run Java_PVR_Flink_Batch PreCommit

**Comment by Lizzfox** [comment]
"Run Website_Stage_GCS PreCommit

**Comment by Amar3tto** [comment]
Run Website_Stage_GCS PreCommit

**Comment by aaltay** [comment]
@aromanenko-dev - do you know anyone who might be able to review this?

**Review by aromanenko-dev — CHANGES_REQUESTED** [review]
Thanks! I left my comments, ptal.

PS: Please, squash multiple commits into one or several atomic ones (if it's needed to have more than one) and avoid merge commits to keep a commit history clean (use `git rebase` instead).

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/package-info.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** Transforms for reading and writing from streaming CDAP plugins. */
+@Experimental(Kind.SOURCE_SINK)
+package org.apache.beam.sdk.io.sparkreceiver;
```
[the remark this answers is not in the document]
Comment (aromanenko-dev) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/package-info.java:21]: Will `SparkReceiverIO` provide a separate user API for read and/or write in this case that is not dependent on `CdapIO`?

**Comment by aromanenko-dev** [comment]
Kind oping on this. What is a state for this PR? Is it ready for another round of review?

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  /**
+   * @return Proxy for given {@param receiver} that doesn't use Spark environment and uses Apache
+   *     Beam mechanisms instead.
+   */
+  public T build()
+      throws InvocationTargetException, InstantiationException, IllegalAccessException {
+
+    if (constructorArgs == null) {
```
[the remark this answers is not in the document]
Comment (Lizzfox) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java]: Done, thanks

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        if (!currArgClass.equals(arg.getClass())) {
+          matches = false;
+          break;
+        }
+      }
+      if (matches) {
+        currentConstructor = constructor;
+      }
+    }
+    if (currentConstructor == null) {
```
[the remark this answers is not in the document]
Comment (Lizzfox) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java]: Done, thanks

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    Constructor<?> currentConstructor = null;
+    for (Constructor<?> constructor : sparkReceiverClass.getDeclaredConstructors()) {
+      Class<?>[] paramTypes = constructor.getParameterTypes();
+      if (paramTypes.length != constructorArgs.length) {
+        continue;
+      }
+      boolean matches = true;
+      for (int i = 0; i < constructorArgs.length; i++) {
+        Object arg = constructorArgs[i];
+        if (arg == null) {
```
[the remark this answers is not in the document]
Comment (Lizzfox) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java]: Done, thanks

**Comment by Lizzfox** [comment]
@aromanenko-dev 
> PS: Please, squash multiple commits into one or several atomic ones (if it's needed to have more than one) and avoid merge commits to keep a commit history clean (use `git rebase` instead).

Could you tell me please what you mean exactly? Do we need to squash the published comments into one or do we need to merge them before pushing them to a remote repository? 

Could the person who will be merging the PR use the `squash and merge` option?

**Comment by Lizzfox** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
@Lizzfox 
> Could you tell me please what you mean exactly? 

You need to rebase your feature branch with `master` branch (e.g. `git checkout <feature-branch> && git rebase`). It will put your commit(s), that you committed in your feature, on top of `master` branch. 

Also, it's usually better to have your local `master` up-to-date with `origin` one (e.g. `git checkout master && git fetch origin && git pull`).

And, please, avoid merge `master` branch into your feature branch, use `git rebase` instead, as I mentioned above.

> Do we need to squash the published comments into one or do we need to merge them before pushing them to a remote repository?

Yes, you need to squash all commits into one or into several logical ones with a proper commit message. That will help to keep commits history more clear and evident. Also, it should be much easier to rollback such changes if needed.

> Could the person who will be merging the PR use the squash and merge option?

Yes, it's possible, but it will squash all your commits into only one and the squashed commits should not contain merge commits (like it has for now, as I can see).

Feel free to ping me on `the-asf` channel on Slack if you have more questions on this!

**Comment by aromanenko-dev** [comment]
@Lizzfox `Java PreCommit` job fails because of this:
```
18:41:01 > Task :sdks:java:io:sparkreceiver:compileJava
18:41:01 /home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_Phrase/src/sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java:84: error: [dereference.of.nullable] dereference of possibly-null reference currentConstructor
18:41:01     currentConstructor.setAccessible(true);
```
Could you fix it?

You can text it locally with this command:
```
./gradlew :sdks:java:io:sparkreceiver:check
```

**Comment by Amar3tto** [comment]
Run Go PreCommit

**Comment by Amar3tto** [comment]
Run Java PreCommit

File: sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import java.lang.reflect.InvocationTargetException;
+import javax.annotation.Nullable;
+import org.apache.commons.lang.ClassUtils;
+import org.apache.spark.streaming.receiver.Receiver;
+
+/**
+ * Class for building an instance for {@link Receiver} that uses Apache Beam mechanisms instead of
+ * Spark environment.
+ */
+@SuppressWarnings({
```
[the remark this answers is not in the document]
Comment (Lizzfox) [reply sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReceiverBuilder.java]: SOLVED: used `checkStateNotNull()` instead. It fixed!

**Comment by aromanenko-dev** [comment]
@Amar3tto Could you rebase your feature branch against `master` and squash your commits? Thanks!

**Comment by Amar3tto** [comment]
> @Amar3tto Could you rebase your feature branch against `master` and squash your commits? Thanks!

I've squashed all commits into one.

**Comment by Amar3tto** [comment]
Run Java PreCommit

**Comment by aromanenko-dev** [comment]
@Amar3tto @Lizzfox 
I'm sorry for delay with a review, I was on holidays. 

Thanks for addressing the comments! It looks ok for me in general but it's still not clear why do we need a new dedicated IO for that since it doesn't provide any IO user API? Could you give more details about the next steps on this IO?

**Review by aromanenko-dev — APPROVED** [review]
LGTM

**Comment by aromanenko-dev** [comment]
@Lizzfox Thanks for details! Feel free to ping me for other PRs when they are ready

**Pull request merged**